engines/twine/redraw.cpp

```diff
@@ -0,0 +1,802 @@
+		else
+			rightValue = currentRedrawList[i].right;
+
+		if (currentRedrawList[i].top >= top)
+			topValue = top;
+		else
+			topValue = currentRedrawList[i].top;
+
+		if (currentRedrawList[i].bottom <= bottom)
+			bottomValue = bottom;
+		else
+			bottomValue = currentRedrawList[i].bottom;
+
+		if ((rightValue - leftValue) * (bottomValue - topValue) < ((currentRedrawList[i].bottom - currentRedrawList[i].top) * (currentRedrawList[i].right - currentRedrawList[i].left) + area)) {
+			currentRedrawList[i].left = leftValue;
+			currentRedrawList[i].top = topValue;
+			currentRedrawList[i].right = rightValue;
+			currentRedrawList[i].bottom = bottomValue;
+
+			if (currentRedrawList[i].bottom >= SCREEN_WIDTH)
+				currentRedrawList[i].bottom = SCREEN_TEXTLIMIT_BOTTOM;
+			return;
+		}
+
+		i++;
+	};
+
+	currentRedrawList[i].left = left;
+	currentRedrawList[i].top = top;
+	currentRedrawList[i].right = right;
+	currentRedrawList[i].bottom = bottom;
+
+	if (currentRedrawList[i].bottom >= SCREEN_WIDTH)
+		currentRedrawList[i].bottom = SCREEN_TEXTLIMIT_BOTTOM;
+
+	numOfRedrawBox++;
+}
+
+void Redraw::addRedrawArea(int32 left, int32 top, int32 right, int32 bottom) {
+	if (left < 0)
+		left = 0;
+	if (top < 0)
+		top = 0;
+	if (right >= SCREEN_WIDTH)
+		right = SCREEN_TEXTLIMIT_RIGHT;
+	if (bottom >= SCREEN_HEIGHT)
+		bottom = SCREEN_TEXTLIMIT_BOTTOM;
+
+	if (left > right || top > bottom)
+		return;
+
+	nextRedrawList[currNumOfRedrawBox].left = left;
+	nextRedrawList[currNumOfRedrawBox].top = top;
+	nextRedrawList[currNumOfRedrawBox].right = right;
+	nextRedrawList[currNumOfRedrawBox].bottom = bottom;
+
+	currNumOfRedrawBox++;
+
+	addRedrawCurrentArea(left, top, right, bottom);
+}
+
+void Redraw::moveNextAreas() {
+	int32 i;
+
+	numOfRedrawBox = 0;
+
+	for (i = 0; i < currNumOfRedrawBox; i++) {
+		addRedrawCurrentArea(nextRedrawList[i].left, nextRedrawList[i].top, nextRedrawList[i].right, nextRedrawList[i].bottom);
+	}
+}
+
+void Redraw::flipRedrawAreas() {
+	int32 i;
+
+	for (i = 0; i < numOfRedrawBox; i++) { // redraw areas on screen
+		copyBlockPhys(currentRedrawList[i].left, currentRedrawList[i].top, currentRedrawList[i].right, currentRedrawList[i].bottom);
+	}
+
+	numOfRedrawBox = 0;
+
+	for (i = 0; i < currNumOfRedrawBox; i++) { //setup the redraw areas for next display
+		addRedrawCurrentArea(nextRedrawList[i].left, nextRedrawList[i].top, nextRedrawList[i].right, nextRedrawList[i].bottom);
+	}
+}
+
+void Redraw::blitBackgroundAreas() {
+	int32 i;
+	const RedrawStruct *currentArea = currentRedrawList;
+
+	for (i = 0; i < numOfRedrawBox; i++) {
+		_engine->_interface->blitBox(currentArea->left, currentArea->top, currentArea->right, currentArea->bottom, (int8 *)_engine->workVideoBuffer, currentArea->left, currentArea->top, (int8 *)_engine->frontVideoBuffer);
+		currentArea++;
+	}
+}
+
+void Redraw::sortDrawingList(DrawListStruct *list, int32 listSize) {
+	int32 i;
+	int32 j;
+
+	DrawListStruct tempStruct;
+
+	for (i = 0; i < listSize - 1; i++) {
```

Comment:
I would write C++ `for (int i = 0;...`

Reply:
Yes, this was c89 code initially - will try to clean stuff up once I get there. There are a lot of other places left.

Reply:
this is fixed now (but still not all locations)